File: leetcode1124/solveLeetcode1124_method2.go
// func longestWPI(hours []int) int {

// 	n := len(hours)

// 	ans, s := 0, 0

// 	pos := make([]int, n+2) // 记录前缀和首次出现的位置

// 	for i := 1; i <= n; i++ {

// 		if hours[i-1] > 8 {

// 			s--

// 		} else {

// 			s++

// 		} // 所有和前缀和相关的都取反

// 		if s < 0 {

// 			ans = i

// 		} else {

// 			if pos[s+1] != 0 {

// 				ans = int(max(ans,i-pos[s+1]))

// 			}

// 			if pos[s] == 0 {

// 				pos[s] = i

// 			}

// 		}

// 	}

// 	return ans

// }

File: DataStructure/prefix/leetcode1124/solveLeetcode1124_method2.cpp
// class Solution {
// public:
//     int longestWPI(vector<int>& hours) 
//     {
//         const int n=hours.size();
//         //printf("********************%d\n",n);
//         //stack<int>st;
//         vector<int>prefix(n+1),st;
//         st.emplace_back(0);


//         int i;
//         for(i=1;i<=n;++i)
//         {
//             //printf("********************\n");

//             prefix[i] = prefix[i-1] + (hours[i-1] > 8 ? 1: -1);

//             if(prefix[i]<prefix[st.back()])st.emplace_back(i);
//         }


//         //printf("********************\n");

//         int ans=0;
//         for(i=n;i;--i)
//         {
//             while(!st.empty() && prefix[i]>prefix[st.back()])
//             {
//                 ans=max(ans,i-st.back());
//                 st.pop_back();
//             }
//         }

//         return ans;
//     }
// };